src/index.test.ts

```diff
@@ -1,22 +1,208 @@
+{"host":"connectorECSTaskId","source":"connector","sourcetype":"ST004:application_json","index":"pay_application","event":"Connector application log line 2","fields":{"account":"test","environment":"test-12"}}`
+
+    expect(result.records[0].result).toEqual('Ok')
+    expect(result.records[0].recordId).toEqual('applicationLogEvent-1')
+    expect(Buffer.from(result.records[0].data, 'base64').toString()).toEqual(expected_decoded_data_1)
+
+const expected_decoded_data_2 = `{"host":"frontendECSTaskId","source":"frontend","sourcetype":"ST004:application_json","index":"pay_application","event":"Frontend application log line 3","fields":{"account":"test","environment":"test-12"}}
+{"host":"frontendECSTaskId","source":"frontend","sourcetype":"ST004:application_json","index":"pay_application","event":"Frontend application log line 4","fields":{"account":"test","environment":"test-12"}}`
 
-    const event = {
-      'records': [
+    expect(result.records[1].result).toEqual('Ok')
+    expect(result.records[1].recordId).toEqual('applicationLogEvent-2')
+    expect(Buffer.from(result.records[1].data, 'base64').toString()).toEqual(expected_decoded_data_2)
+  })
+
+  test('should drop CloudWatch logs which are not DATA_MESSAGE', async () => {
+    const result = await handler(anApplicationLogEvent)
+    expect(result.records[2].result).toEqual('Dropped')
+  })
+
+  test('should include record id in errors', () => {
+    const unprocessableAppLog: FirehoseTransformationEvent = {
+      deliveryStreamArn: 'someDeliveryStreamArn',
+      invocationId: 'someId',
+      region: 'eu-west-1',
+      records: [
         {
-          'recordId': 'a record id',
-          'approximateArrivalTimestamp': 1702212345678,
-          'data': Buffer.from(JSON.stringify('hellow world')).toString('base64'),
-        },
-      ],
+          approximateArrivalTimestamp: 1234,
+          recordId: 'unprocessableAppLog-1',
+          data: Buffer.from(JSON.stringify(
+              {
+                "owner": "223851549868",
+                "logGroup": "test-12_UNKNOWN_connector",
+                "logStream": "connectorECSTaskId",
+                "subscriptionFilters": [],
+                "messageType": "DATA_MESSAGE",
+                "logEvents": [
+                    {
+                        "id": "cloudwatch-log-message-id-1",
+                        "timestamp": "1234",
+                        "message": "Connector application log line 1"
+                    },
+                    {
+                        "id": "cloudwatch-log-gmessage-id-2",
+                        "timestamp": "12345",
+                        "message": "Connector application log line 2"
+                    }
+                ]
+            })
+          ).toString('base64')
+        }]
+      }
+    const expectedErrorMessage = 'Error processing record "unprocessableAppLog-1": Unknown host for log type of "UNKNOWN"'
```

Comment:
Can you also add tests for unhappy paths for source_type, source, and index?

Reply:
Yes I have restructured things a bit to improve checking and error handling.